src/server/search_family.cc

```diff
@@ -0,0 +1,239 @@
+// Copyright 2022, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "server/search_family.h"
+
+#include <variant>
+#include <vector>
+
+#include "base/logging.h"
+#include "core/search/search.h"
+#include "facade/reply_builder.h"
+#include "server/command_registry.h"
+#include "server/conn_context.h"
+#include "server/container_utils.h"
+#include "server/engine_shard_set.h"
+#include "server/transaction.h"
+
+extern "C" {
+#include "redis/listpack.h"
+#include "redis/object.h"
+};
+
+namespace dfly {
+
+using namespace std;
+using namespace facade;
+
+namespace {
+
+string_view SdsToSafeSv(sds str) {
+  return str != nullptr ? string_view{str, sdslen(str)} : ""sv;
+}
+
+using DocumentData = absl::flat_hash_map<std::string, std::string>;
+using SerializedDocument = pair<std::string /*key*/, DocumentData>;
+using Query = search::AstExpr;
+
+struct BaseAccessor : public search::HSetAccessor {
+  using FieldConsumer = search::HSetAccessor::FieldConsumer;
+
+  virtual DocumentData Serialize() const = 0;
+};
+
+struct ListPackAccessor : public BaseAccessor {
+  using LpPtr = uint8_t*;
+
+  ListPackAccessor(LpPtr ptr) : lp_{ptr} {
+  }
+
+  bool Check(FieldConsumer f, string_view active_field) const override {
+    uint8_t intbuf[LP_INTBUF_SIZE];
+
+    if (!active_field.empty()) {
+      return f(container_utils::LpFind(lp_, active_field, intbuf).value_or(""));
+    }
+
+    uint8_t* fptr = lpFirst(lp_);
+    while (fptr) {
+      fptr = lpNext(lp_, fptr);  // skip key
```

Comment:
Perhaps `DCHECK()` that `fptr != nullptr`? If you choose to do so, please also add one in `Serialize()` below